emscripten.py

```diff
@@ -1480,10 +1480,27 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
     receiving = '\n'.join(['var real_' + s + ' = asm["' + s + '"]; asm["' + s + '''"] = function() {''' + runtime_assertions + '''  return real_''' + s + '''.apply(null, arguments);
 };
 ''' for s in exported_implemented_functions if s not in ['_memcpy', '_memset', 'runPostSets', '_emscripten_replace_memory', '__start_module']])
+
+  module_exports = exported_implemented_functions + function_tables(function_table_data)
+
+  # Generate an exports file that records all the exported symbols from asm.js/wasm module.
+  module_exports_suppressions = '\n'.join(['/**\n * @suppress {duplicate, undefinedVars}\n */\nvar %s;\n' % i for i in module_exports])
+  exports_file = get_configuration().get_temp_files().get('_module_exports.js')
+  shared.Settings.MODULE_EXPORTS_FILE = exports_file.name
+  exports_file.write(module_exports_suppressions.encode())
+  exports_file.close()
+
   if not shared.Settings.SWAPPABLE_ASM_MODULE:
```

Comment:
trying to check out what happens with metadce, I noticed that since we actually use `SWAPPABLE_ASM_MODULE` by default for wasm (for async compilation), so this optimization is not occurring for wasm anyhow (except when async compilation is off)

Reply:
That looks right :/ We can't do this for `SWAPPABLE_ASM_MODULE` it seems. Disabled it for that scenario for now, so when `SWAPPABLE_ASM_MODULE`, then write out the exports in full.

Reply:
I missed something - why can't we do it for the swappable case? Might be more complicated to add the indirection, but not impossible, I'd hope, but I didn't try.

if we can't do this for the swappable case, then probably we should not do it for the wasm case, since wasm+sync compilation is not an important mode.